ykllvmwrap/src/ykllvmwrap.cc

```diff
@@ -394,24 +398,20 @@ extern "C" void *__ykllvmwrap_irtrace_compile(char *FuncNames[], size_t BBs[],
       }
 
       // If execution reaches here, then the instruction I is to be copied into
-      // JITMod.
-      //
-      // FIXME: We now scan the instruction's operands checking that each is
-      // defined in JITMod. Any variable not defined means that the
-      // corresponding variable in AOTMod was instantiated prior to tracing,
-      // but was not marked as a trace input. Currently this only applies to
-      // variables relating to starting and stopping tracing. For those
-      // variables we simply allocate dummy storage so that the module will
-      // verify and compile. In the long run we should omit these instructions
-      // as tracing them is never beneficial. Then the below hack can be
-      // removed.
+      // JITMod. Before we can do this, we have to scan the instruction's
+      // operands checking that each is defined in JITMod.
       for (unsigned OpIdx = 0; OpIdx < I->getNumOperands(); OpIdx++) {
         Value *Op = I->getOperand(OpIdx);
         if (VMap[Op] == nullptr) {
-          // Value is undefined in JITMod.
+          // The operand is undefined in JITMod.
           Type *OpTy = Op->getType();
           if (isa<llvm::AllocaInst>(Op)) {
-            // Value is a stack allocation, so make a dummy stack slot.
+            // The operand is alocated on the stack with an `alloca`.
+            //
+            // LLVM has a tendency to move allocas up to the first block of a
+            // function, and if we didn't trace that block, then we will have
+            // missed those allocations. In these cases we materialise the
+            // allocations as we see them used in code that *was* traced.
```

Comment:
Does this mean that we duplicate some allocas if they were included in the trace?

Reply:
No. Each variable has exactly one alloca.

Reply:
How do we guarantee that?

Reply:
The LLVM verifier would have rejected the module at AOT compile-time.

Reply:
And in fact, since each alloca is a pointer address, they can't be duplicated.

Reply:
Let me change my question: do we explicitly check "ah, this var already has an alloca in the trace so we don't need to add another"?

Reply:
Indirectly, yes. We only reach this code if the variable is [as-yet undefined](https://github.com/ykjit/yk/blob/4d6276ced71842c0495379e3763c5cdf17161b4e/ykllvmwrap/src/ykllvmwrap.cc#L405-L416).

Reply:
OK, maybe we should add that to the comment (or, even better, add an assert along those lines, if that's possible)?

Reply:
Well, we can, but the line where we do the check is only a few lines above. Is it really necessary?

Reply:
[this one](https://github.com/ykjit/yk/blob/4d6276ced71842c0495379e3763c5cdf17161b4e/ykllvmwrap/src/ykllvmwrap.cc#L405)

Reply:
@ltratt ping

Reply:
OK, I see. I think a more clear comment that this is what happening at line 405ish would really help (at least it would help me).

Reply:
The comment at line 406 is about as good as it gets:

> // The operand is undefined in JITMod.

But maybe it helps to reiterate this a few lines later? Try b59cf77008?